File: 2021/day01-2.hs
import System.IO

solution :: String -> Int
solution content = length . filter id . isIncreasing . windows . map read . lines $ content

windows :: [Int] -> [Int]
windows (x:y:z:[]) = [x+y+z]
windows (x:y:z:xs) = [x+y+z] ++ (windows ([y, z] ++ xs))

isIncreasing :: [Int] -> [Bool]
isIncreasing (x:y:[]) = [y > x]
isIncreasing (x:y:xs) = [y > x] ++ (isIncreasing ([y] ++ xs))

main :: IO ()
main = do
    handle <- openFile "day01.dat" ReadMode
    content <- hGetContents handle
    print . solution $ content

File: 2021/day01-2.ts
function solution(content: string) {
  const depths = content.split("\n").map((item) => parseInt(item));
  const windows = depths
    .slice(2)
    .map((item, index) => item + depths[index] + depths[index + 1]);
  return windows
    .slice(1)
    .map((item, index) => item > windows[index])
    .filter((item) => item).length;
}

Deno.readTextFile("day01.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
